cartographer/mapping_3d/sparse_pose_graph/optimization_problem.cc

```diff
@@ -146,58 +153,67 @@ void OptimizationProblem::Solve(
         C_point_clouds[constraint.j].translation());
   }
 
-  CHECK(!node_data_.empty());
-  CHECK_GE(trajectories.size(), node_data_.size());
-
-  // Add constraints for IMU observed data: angular velocities and
-  // accelerations.
-  auto it = imu_data_.cbegin();
-  while ((it + 1) != imu_data_.cend() && (it + 1)->time <= node_data_[0].time) {
-    ++it;
-  }
+  for (const auto& trajectory_nodes_pair : nodes_per_trajectory) {
+    const mapping::Submaps* const trajectory = trajectory_nodes_pair.first;
+    const std::vector<size_t>& node_indices = trajectory_nodes_pair.second;
+    const std::deque<ImuData>& imu_data = imu_data_.at(trajectory);
+    CHECK(!node_indices.empty());
+    CHECK(!imu_data.empty());
+
+    // Add constraints for IMU observed data: angular velocities and
```

Comment:
Add constraints based on IMU observations of angular velocity and linear acceleration.

Also, add a comment about this first loop?